tests/resp.py

```diff
@@ -0,0 +1,532 @@
+            if code == b"|":
+                yield Attribute(result_map), rest
+            yield result_map, rest
+
+        elif code == b"-":  # error
+            # we decode them automatically
+            decoded = self.decode_bytes(arg)
+            assert isinstance(decoded, str)
+            code, value = decoded.split(" ", 1)
+            yield ErrorStr(code, value), rest
+
+        elif code == b"!":  # resp3 error
+            count = int(arg)
+            expect = count + 2  # +2 for the trailing CRNL
+            while len(rest) < expect:
+                incoming = yield (None)
+                assert incoming is not None
+                rest += incoming
+            bulkstr = rest[:count]
+            decoded = self.decode_bytes(bulkstr)
+            assert isinstance(decoded, str)
+            code, value = decoded.split(" ", 1)
+            yield ErrorStr(code, value), rest[expect:]
+
+        else:
+            raise ValueError(f"Unknown opcode '{code.decode()}'")
+
+
+class NeedMoreData(RuntimeError):
+    """
+    Raised when more data is needed to complete a parse
+    """
+
+
+class RespParser:
+    """
+    A class for simple RESP protocol decoding for unit tests.
+    Uses a RespGeneratorParser to produce data, and can
+    produce top-level objects for as long as there is data available.
+    """
+
+    def __init__(self) -> None:
+        self.parser = RespGeneratorParser()
+        self.generator: Optional[
+            Generator[Optional[Tuple[Any, bytes]], Union[None, bytes], None]
+        ] = None
+        # which has not resulted in a parsed value
+        self.consumed: List[bytes] = []
+
+    def parse(self, buffer: bytes) -> Optional[Any]:
+        """
+        Parse a buffer of data, return a tuple of a single top-level primitive and the
+        remaining buffer or raise NeedMoreData if more data is needed to
+        produce a value.
+        """
+        if self.generator is None:
+            # create a new parser generator, initializing it with
+            # any unparsed data from previous calls
+            buffer = b"".join(self.consumed) + buffer
+            del self.consumed[:]
+            self.generator = self.parser.parse(buffer)
+            parsed = self.generator.send(None)
+        else:
+            # sen more data to the parser
+            parsed = self.generator.send(buffer)
+
+        if parsed is None:
+            self.consumed.append(buffer)
+            raise NeedMoreData()
+
+        # got a value, close the generator, store the remaining buffer
+        self.generator.close()
+        self.generator = None
+        value, remaining = parsed
+        self.consumed = [remaining]
+        return value
+
+    def get_unparsed(self) -> bytes:
+        return b"".join(self.consumed)
+
+    def close(self) -> None:
+        if self.generator is not None:
+            self.generator.close()
+            self.generator = None
+        del self.consumed[:]
```

Comment:
maybe do `self.consumed.clear()` here to make it more readable? same for all other places

Reply:
fair point